wgengine/router/router_windows.go

```diff
@@ -160,3 +107,83 @@ func (r *winRouter) Close() error {
 func cleanup(logf logger.Logf, interfaceName string) {
 	// Nothing to do here.
 }
+
+// firewallTweaker changes the Windows firewall. Normally this wouldn't be so complicated,
+// but it can be REALLY SLOW to change the Windows firewall for reasons not understood.
+// Like 4 minutes slow. But usually it's tens of milliseconds.
+// See https://github.com/tailscale/tailscale/issues/785.
+// So this tracks the desired state and runs the actual adjusting code asynchrounsly.
+type firewallTweaker struct {
+	logf logger.Logf
+
+	mu      sync.Mutex
+	running bool   // doAsyncSet goroutine is running
+	known   bool   // firewall is in known state (in lastVal)
+	want    string // next value we want, or "" to delete the firewall rule
+	lastVal string // last set value, if known
+}
+
+func (ft *firewallTweaker) clear() { ft.set("") }
+
+func (ft *firewallTweaker) set(ip string) {
+	ft.mu.Lock()
+	defer ft.mu.Unlock()
+
+	if ip == "" {
+		ft.logf("marking for removal")
+	} else {
+		ft.logf("marking desired IP %v", ip)
+	}
+	ft.want = ip
+	if ft.running {
+		// The doAsyncSet goroutine will check ft.want
+		// before returning.
+		return
+	}
+	ft.running = true
+	go ft.doAsyncSet()
+}
+
+func (ft *firewallTweaker) doAsyncSet() {
+	bo := backoff.NewBackoff("win-firewall", ft.logf, time.Minute)
+	ctx := context.Background()
+
+	ft.mu.Lock()
+	for { // invariant: ft.mu must be locked when beginning this block
+		val := ft.want
+		if ft.known && ft.lastVal == val {
+			ft.running = false
+			ft.mu.Unlock()
+			return
+		}
+		ft.mu.Unlock()
+
+		var cmd *exec.Cmd
+		t0 := time.Now()
+		if val == "" {
+			ft.logf("deleting Tailscale-In firewall rule")
+			cmd = exec.Command("netsh", "advfirewall", "firewall", "delete", "rule", "name=Tailscale-In", "dir=in")
+		} else {
+			ft.logf("setting Tailscale-In firewall IP to %q", val)
+			cmd = exec.Command("netsh", "advfirewall", "firewall", "add", "rule", "name=Tailscale-In", "dir=in", "action=allow", "localip="+val, "profile=private", "enable=yes")
```

Comment:
This correctly replaces a rule that already exists? No need to go through a delete+add?

Reply:
Whoops. I totally misremembered how it worked. I thought the `name=` was a primary key, but nope:

```
Y:\>netsh advfirewall firewall add rule name=Foo dir=in action=allow localip=1.2.4.5 enable=yes
Ok.


Y:\>netsh advfirewall firewall add rule name=Foo dir=in action=allow localip=1.2.4.6 enable=yes
Ok.


Y:\>netsh advfirewall firewall show rule name=Foo

Rule Name:                            Foo
----------------------------------------------------------------------
Enabled:                              Yes
Direction:                            In
Profiles:                             Domain,Private,Public
Grouping:
LocalIP:                              1.2.4.6/32
RemoteIP:                             Any
Protocol:                             Any
Edge traversal:                       No
Action:                               Allow

Rule Name:                            Foo
----------------------------------------------------------------------
Enabled:                              Yes
Direction:                            In
Profiles:                             Domain,Private,Public
Grouping:
LocalIP:                              1.2.4.5/32
RemoteIP:                             Any
Protocol:                             Any
Edge traversal:                       No
Action:                               Allow
Ok.


Y:\>netsh advfirewall firewall delete rule name=Foo

Deleted 2 rule(s).
Ok.
```

So this'll need a bit of changing.

Reply:
But I guess that's good for IPv6 support.

Reply:
I don't see you change code as @danderson suggested. I guess it is still comming.